selenium/src/main/java/com/deque/html/axecore/selenium/AxeBuilder.java

```diff
@@ -686,32 +706,53 @@ private ArrayList<String> runPartialRecursive(
           (String)
               WebDriverInjectorExtensions.executeAsyncScript(
-                  webDriver, runPartialScript, context, options);
-      ArrayList<String> partialResults = new ArrayList<String>();
+                  webDriver, runPartialScript, context, options, frameStack);
       partialResults.add(resResponse);
       if (disableIframeTesting) {
         return partialResults;
       }
 
       for (FrameContext fc : contexts) {
-        Object frameContext = AxeReporter.serialize(fc.getFrameContext());
-        Object frameSelector = AxeReporter.serialize(fc.getFrameSelector());
-        Object frame =
-            WebDriverInjectorExtensions.executeScript(webDriver, shadowSelectScript, frameSelector);
-
-        if (frame instanceof String) {
-          webDriver.switchTo().frame((String) frame);
-        } else if (frame instanceof WebElement) {
-          webDriver.switchTo().frame((WebElement) frame);
-        } else if (frame instanceof Integer) {
-          webDriver.switchTo().frame((Integer) frame);
-        } else {
+        try {
+          Object frameContext = AxeReporter.serialize(fc.getFrameContext());
+          Object frameSelector = AxeReporter.serialize(fc.getFrameSelector());
+          Object frame =
+              WebDriverInjectorExtensions.executeScript(
+                  webDriver, shadowSelectScript, frameSelector);
+
+          if (frame instanceof String) {
+            webDriver.switchTo().frame((String) frame);
+          } else if (frame instanceof WebElement) {
+            webDriver.switchTo().frame((WebElement) frame);
+          } else if (frame instanceof Integer) {
+            webDriver.switchTo().frame((Integer) frame);
+          } else {
+            partialResults.add(null);
+            continue;
+          }
+          frameStack.push(frameSelector);
+
+          ArrayList<String> morePartialResults =
+              runPartialRecursive(webDriver, options, frameContext, false, frameStack);
+          partialResults.addAll(morePartialResults);
+        } catch (org.openqa.selenium.TimeoutException e) {
+          webDriver.switchTo().window(windowHandle);
+          for (Object frameSelector : frameStack) {
+            Object frame =
+                WebDriverInjectorExtensions.executeScript(
+                    webDriver, shadowSelectScript, frameSelector);
+            if (frame instanceof String) {
+              webDriver.switchTo().frame((String) frame);
+            } else if (frame instanceof WebElement) {
+              webDriver.switchTo().frame((WebElement) frame);
+            } else if (frame instanceof Integer) {
+              webDriver.switchTo().frame((Integer) frame);
+            }
+          }
           partialResults.add(null);
           continue;
         }
-        ArrayList<String> morePartialResults =
-            runPartialRecursive(webDriver, options, frameContext, false);
-        partialResults.addAll(morePartialResults);
+        frameStack.pop();

```

Comment:
Shouldn't need this anymore, and in fact this would probably cause an issue since you're popping the parent frame before all children frame have been processed:

For example, if you had this as the frame stack `['ancestor', 'parent']` and are currently under `parent` frame. If parent frame had 2 subframes and the first one is unloaded, you would switch to the parent frame in the catch statement but then pop parent off the frame stack and would have a bad frame stack when trying to switch to the 2nd child.

Reply:
The catch statement ends with a `continue`. The `pop` only occurs if we get to the end of the loop without exceptions